skfp/utils/splitters.py

```diff
@@ -0,0 +1,630 @@
+import numbers
+import warnings
+from collections import defaultdict
+from collections.abc import Sequence
+from itertools import chain
+from typing import Any, Optional, Union
+
+import numpy as np
+import sklearn.utils as skutils
+from rdkit.Chem import Mol
+from rdkit.Chem.Scaffolds import MurckoScaffold
+
+from skfp.utils.validators import Interval, RealNotInt, ensure_mols, validate_params
+
+
+@validate_params(
+    {
+        "data": ["sequence"],
+        "additional_data": ["list"],
+        "train_size": [
+            Interval(RealNotInt, 0, 1, closed="neither"),
+            Interval(numbers.Integral, 1, float("inf"), closed="left"),
+            None,
+        ],
+        "test_size": [
+            Interval(RealNotInt, 0, 1, closed="neither"),
+            Interval(numbers.Integral, 1, float("inf"), closed="left"),
+            None,
+        ],
+        "include_chirality": ["boolean"],
+        "return_indices": ["boolean"],
+    }
+)
+def scaffold_train_test_split(
+    data: Sequence[Union[str, Mol]],
+    *additional_data: list[Sequence],
+    train_size: Optional[float] = None,
+    test_size: Optional[float] = None,
+    include_chirality: bool = False,
+    return_indices: bool = False,
+) -> Union[
+    tuple[list[Union[str, Mol]], list[Union[str, Mol]], list[Sequence[Any]]],
+    tuple[list[int], list[int], list[Sequence[Any]]],
+    tuple[list[int], list[int]],
+]:
+    """
+    Split a list of SMILES or RDKit `Mol` objects into train and test subsets using Bemis-Murcko [1]_ scaffolds.
+
+    This function ensures that similar molecules (sharing the same scaffold) are in the same split.
+    This method, suggested in the MoleculeNet [2]_ paper, helps to test the model's ability to
+    generalize to entirely new scaffolds. MoleculeNet introduced the scaffold split as an approximation
+    to the time split, assuming that new molecules (test set) will be structurally different in terms of
+    scaffolds from the training set.
+
```

Comment:
Maybe add a description about known limitations, e.g. disconnected molecules and those without rings get no scaffold and are put into one group regardless of their structure. Same for train-valid-test docs.